### DISCUSSION TITLE
Building a caching layer

### DISCUSSION THREAD

**Discussion by varunbpatil** [benbjohnson/wtf#40] [Q&A] [answered]
Hi @benbjohnson,

First of all, many thanks for creating this repository. I'm a Go newbie and I'm learning a lot from this project alone.

My question is based on this article - https://www.gobeyond.dev/standard-package-layout/ - where you give an example of a caching layer on top of PostgreSQL.

```Go
package myapp

// UserCache wraps a UserService to provide an in-memory cache.
type UserCache struct {
        cache   map[int]*User
        service UserService
}

...
```

From the looks of it, this is supposed to go into the root package (inferring from the `package myapp`) presumably because it has no dependency apart from the application domain itself. 

But, isn't `map[int]*User` an implementation detail? What if my application needs to run on multiple nodes and I need to use Redis or Memcached for caching? Now, I'll be needing to make changes to the application domain!

Also, since `UserCache` doesn't actually implement the `UserService` interface, I'll need to use `(*UserCache).User()` for one thing and `(*UserService).CreateUser()` and `(*UserService).DeleteUser()` for other things which doesn't seem very nice (_Note to other readers_ - the names of the interface methods has changed in wtfdial, but is still relevant).

I was thinking of two ways to approach this:

1. Implement the `UserService` interface inside `inmem/user.go` (for in-memory caching using map).  All methods would just call the corresponding method in `UserService` which would be injected as a dependency. Only the `User()` method would actually do caching. Similarly, I would have `memcached/user.go` or `redis/user.go` if I need to.
2. Create a `wtf.CacheService` interface in the root package with `get()` and `set()` methods and then inject a particular cache implementation (inmem/redis/memcache) into `sqlite.DB` just like you have already done for `wtf.EventService`.

 Would like to know your thoughts on this.

**Comment by benbjohnson** [comment, marked as answer]
@varunbpatil I'm glad to hear the repo has been helpful. I'll try to answer your questions.

> _From the looks of it, this is supposed to go into the root package_

Yes, I put it in the domain package because it has no other dependencies, however, I could also see having an `inmem` package and moving it there.

> _Also, since UserCache doesn't actually implement the UserService interface_

The `UserCache` should implement the `UserService` as you mentioned in your second approach. That's the way I would go.

> _All methods would just call the corresponding method in UserService which would be injected as a dependency. Only the User() method would actually do caching._

Yes, you can populate the cache through `User()` although you would need to invalidate the cache entries when they're updated or deleted.

> _Similarly, I would have memcached/user.go or redis/user.go if I need to._

👍 

> _Create a wtf.CacheService interface in the root package with get() and set() methods and then inject a particular cache implementation_

I'd personally avoid having a generic cache implementation since you'd need to get/set using `interface{}` and convert between that and the implementation type. Generics in Go should make that better. Also, you may have different rules about how data is invalidated from the cache (e.g. updating a user may invalidate other types of data). That being said, if you have a lot of caches then a generic one might make sense.

**Reply by varunbpatil** [reply]
Thank you very much.
